### DISCUSSION TITLE
advice in dealing with clang bug: use non-default prefix?

### DISCUSSION THREAD

**Discussion by jaysachs** [Homebrew/discussions#5642] [Everyday usage] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.3.24
ORIGIN: https://github.com/Homebrew/brew
HEAD: 916044581862c32fc2365e8e9ff0b1507a98925e
Last commit: 6 days ago
Core tap JSON: 24 Sep 22:07 UTC
Core cask tap JSON: 24 Sep 22:07 UTC
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 16
Homebrew Ruby: 3.3.4 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/3.3.4_1/bin/ruby
CPU: 16-core 64-bit kabylake
Clang: 16.0.0 build 1600
Git: 2.39.5 => /Applications/Xcode.app/Contents/Developer/usr/bin/git
Curl: 8.7.1 => /usr/bin/curl
macOS: 14.6.1-x86_64
CLT: 15.3.0.0.1.1708646388
Xcode: 16.0
```

### Output of `brew doctor`

```shell
Warning: Some installed casks are deprecated or disabled.
You should find replacements for the following casks:
  wkhtmltopdf

Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  openssl@1.1

Warning: Unbrewed dylibs were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae and may need to be deleted.

Unexpected dylibs:
  /usr/local/lib/libdvdcss.2.dylib
  /usr/local/lib/libwkhtmltox.0.12.6.dylib

Warning: Unbrewed header files were found in /usr/local/include.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae and may need to be deleted.

Unexpected header files:
  /usr/local/include/wkhtmltox/image.h
  /usr/local/include/wkhtmltox/pdf.h

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  python@3.12

Warning: Homebrew's "sbin" was not found in your PATH but you have installed
formulae that put executables in /usr/local/sbin.
Consider setting your PATH for example like so:
  echo 'export PATH="/usr/local/sbin:$PATH"' >> ~/.zshrc
```

### Description of issue

TLDR: how weight-bearing are the fairly dire warnings about installing Homebrew outside of the default prefix?

Background: to the best of my understanding, it looks like `/usr/bin/clang` in XCode 16 (and possibly others) has a bug, in that `-isystem` directories are not prepended to the system search list, but instead added *after* `/usr/local/include`. This is a problem for some builds depending on e.g. `boringssl`  (under `bazel`) where  the openssl headers installed by homebrew in `/usr/local/include` are found before the boringssl headers (whose location is added via `-isystem`). (I reported the `/usr/bin/clang` bug [here](https://feedbackassistant.apple.com/feedback/15286196)). 

If someone has another suggestion on how to approach working around this, I'm all ears.

**Comment by carlocab** [comment, marked as answer]
You can `brew unlink openssl@3` to remove the OpenSSL headers from `/usr/local` before building boringssl.

**Reply by jaysachs** [reply]
Interesting. And the consequences of that would be restricted to dealing with non-homebrew builds that I want to depend on the brew-installed openssl? (in which case I'd simply need to ensure `/usr/local/Cellar/openssl@3/3.3.1/include` is in the search path for those compilations) In particular, `unlink`ing would not affect any homebrew-managed libraries/binaries depending on openssl? Or might it affect things that brew needs to build as well?

**Reply by carlocab** [reply]
It could affect things that `brew` needs to build. I suggest doing `brew link openssl@3` when you're done.

**Reply by carlocab** [reply]
> (in which case I'd simply need to ensure `/usr/local/Cellar/openssl@3/3.3.1/include` is in the search path for those compilations)

I also suggest treating `*/Cellar/*` paths as implementation details that don't exist. Instead, use `"$(brew --prefix openssl@3)"` or `/usr/local/opt/openssl@3`.

**Reply by jaysachs** [reply]
Thanks -- after `brew unlink` for openssl, snappy, and leveldb, I can get my bazel build to succeed (or at least past the homebrew-related issues). 

Maybe for a workaround I'll write a wrapper for `brew` that re-links those libraries, does the brew stuff, and then unlinks them ... or, I write a bazel wrapper which unlinks them, runs bazel, and then re-links. Neither is particularly satisfactory, but I run `brew` much less frequently.
